lib/CodeGen/FuseInplaceValue.cpp

```diff
@@ -0,0 +1,76 @@
+//===- FuseInplaceValue.cpp --------------------------------------------------===//
+//
+//                             The ONNC Project
+//
+// See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+#include <onnc/CodeGen/FuseInplaceValue.h>
+#include <onnc/Core/PassAnalysisSupport.h>
+#include <onnc/Core/PassSupport.h>
+
+using namespace onnc;
+
+//===----------------------------------------------------------------------===//
+// FuseInplaceValue
+//===----------------------------------------------------------------------===//
+Pass::ReturnType FuseInplaceValue::runOnModule(Module& pModule)
+{
+  Pass::ReturnType ret = Pass::kModuleNoChanged;
+  Module::cg_iterator cg, cgEnd = pModule.cgEnd();
+  for (cg = pModule.cgBegin(); cg != cgEnd; ++cg)
+    ret |= runOnComputeGraph(*cg->value());
+
+  return ret;
+}
+
+Pass::ReturnType FuseInplaceValue::runOnComputeGraph(ComputeGraph& pCG)
+{
+  Pass::ReturnType ret = Pass::kModuleNoChanged;
+  ComputeGraph::iterator nodeIt, nEnd = pCG.end();
+  for (nodeIt = pCG.begin(); nodeIt != nEnd; ++nodeIt) {
+    ComputeOperator* node = nodeIt;
+    if (!m_IsFusibleFn(*node))
+      continue;
+
+    // TODO: Get fusible input/output pair (backend should provide this)
+    //       Currently assume operator's 1st input and 1st output is a fusible
+    //       pair.
+    const unsigned outputId = 0;
+    Value *input = node->getInput(0),
+          *output = node->getOutput(outputId);
+
+    // If input or output has more than one user, we can't fuse this pair.
+    if (input->getUses().size() > 1 ||
+        output->getUses().size() > 1)
+      continue;
+
+    Define* origDef = input->getDefine();
+    unsigned origDefNo = input->getDefineNo();
+    input->clearDefine();
+    node->replaceOutput(outputId, *input);
+    input->clearDefine();
```

Comment:
This is unnecessary?

Reply:
We want to keep input's original definer so we don't need to modify liveness pass.
e.g. valueA -> OpA -> valueB -> OpB -> valueC -> OpC
And value A, B, C can be fused
=> valueA -> OpA -> valueA -> OpB -> valueA ->OpC
If we stick valueA's definer to original definer, i.e. OpA, then valueA's liveness is [OpA, OpC], but if we don't do it, valueA's definer will be OpC, the liveness will be invalid.

Reply:
But it doesn't have to clear define. It will set to origDefNo later.

Reply:
It is because replaceOutput(unsigned int pIdx, onnc::Value &pValue) need pValue has no definer, so before call replaceOutput(), we need clear pValue's definer.

So do we want to change replaceOutput()'s assumption? Or there is better flow?

Reply:
Oh, it is setDefine()'s assertion. I think this is good, no need to change. 👍

Reply:
Right! OK, thanks : )